internal/runtime/registry/registry.go

```diff
@@ -1,55 +1,224 @@
+	extensions.lock.RLock()
+	defer extensions.lock.RUnlock()
+
+	return extensions.ready
 }
 
-func (r registry) RegisterRuntimeExtension(ext *runtimev1.Extension) {
-	panic("implement me")
+// Add all the Runtime Extensions served by a registered runtime extension server.
+// Please note that if the provided registration object already exists, all the registered
+// RuntimeExtensions gets updated according to the newly provide status.
+func (extensions *extensionRegistry) Add(ext *runtimev1.Extension) error {
+	if !extensions.ready {
+		return errors.New("invalid operation: Get cannot called on a registry not yet ready")
+	}
+
+	if ext == nil {
+		return errors.New("invalid argument, when calling Add ext must not be nil")
+	}
+
+	extensions.lock.Lock()
+	defer extensions.lock.Unlock()
+	return extensions.add(ext)
 }
 
-func (r registry) RemoveRuntimeExtension(ext *runtimev1.Extension) {
-	panic("implement me")
+// Remove all the Runtime Extensions served by a registered runtime extension server.
+func (extensions *extensionRegistry) Remove(ext *runtimev1.Extension) error {
+	if !extensions.ready {
+		return errors.New("invalid operation: Get cannot called on a registry not yet ready")
+	}
+
+	if ext == nil {
+		return errors.New("invalid argument, when calling Remove ext must not be nil")
+	}
+
+	extensions.lock.Lock()
+	defer extensions.lock.Unlock()
+
+	extensions.remove(ext)
+	return nil
 }
 
-func (r registry) GetRuntimeExtension(gvh catalog.GroupVersionHook, name string) RuntimeExtensionRegistration {
-	panic("implement me")
+func (extensions *extensionRegistry) remove(ext *runtimev1.Extension) {
+	for _, e := range extensions.items {
+		if e.RegistrationName == ext.Name {
+			delete(extensions.items, e.Name)
+		}
+	}
 }
 
-func (r registry) GetRuntimeExtensions(gvh catalog.GroupVersionHook) []RuntimeExtensionRegistration {
-	panic("implement me")
+// List all the registered Runtime Extensions for a given Group/Hook..
+func (extensions *extensionRegistry) List(gh catalog.GroupVersionHook) ([]*RuntimeExtensionRegistration, error) {
+	if !extensions.ready {
+		return nil, errors.New("invalid operation: Get cannot called on a registry not yet ready")
+	}
+
+	extensions.lock.RLock()
+	defer extensions.lock.RUnlock()
+
+	l := make([]*RuntimeExtensionRegistration, 0, 10)
+	for _, r := range extensions.items {
+		if r.GroupVersionHook.Group == gh.Group && r.GroupVersionHook.Hook == gh.Hook {
+			l = append(l, r)
+		}
+	}
+	return l, nil
 }
 
-type Registry interface {
-	RemoveRuntimeExtension(ext *runtimev1.Extension)
-	RegisterRuntimeExtension(ext *runtimev1.Extension)
+// Get the Runtime Extensions with a given name.
+func (extensions *extensionRegistry) Get(name string) (*RuntimeExtensionRegistration, error) {
+	if !extensions.ready {
+		return nil, errors.New("invalid operation: Get cannot called on a registry not yet ready")
+	}
+
+	extensions.lock.RLock()
+	defer extensions.lock.RUnlock()
+
+	r, _ := extensions.items[name]
+	return r, nil
```

Comment:
I think it would be good to return an error if we don't find the RuntimeExtensionRegistration in the map.

I think it's more intuitive to get an error in that case vs. having to handle `nil, nil`.

Reply:
I'm not sure about this, this allows to determine not exists without being forced to check for an error type

Reply:
Hm good point.

On one side yes on the other side if I want to get a RuntimeExtensionRegistration with a specific name and it doesn't exist I would usually expect an error.

I think with the current version we have to check `!= error` and `res == nil` and then return the corresponding error. If we just construct the errors correctly here we only need the error check and then propagate that further up.

I assume our caller code doesn't care about if we didn't fine the extension because the registry was not warmed up or because it was never added to the registry.

But I can live with both, up to you